### PR TITLE
New asset handling

### PR DESCRIPTION
**Pull request by StarArawn**
WIP

### CODE REVIEW

File: src/assets/image_asset_manager.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    pub fn update(&mut self, device: &wgpu::Device) -> wgpu::CommandBuffer {
+        let mut encoder = device.create_command_encoder(&wgpu::CommandEncoderDescriptor {
+            label: Some("ImageUpload")
+        });
+
+        for _ in 0..self.image_path_queue.len() {
+            let path = self.image_path_queue.pop().unwrap();
+            let id = self.image_info_manager.insert(PathBuf::from(path.clone()));
+            self.image_info_manager.load(id).unwrap();
+            self.image_info_queue.push((path.clone(), id)).unwrap();
```
[the remark this answers is not in the document]
Comment (floatingmountain) [reply src/assets/image_asset_manager.rs:63]: just pushed get_loaded_once

File: src/assets/image_asset_manager.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            let item = image_info_still_loading.pop().unwrap();
+            self.image_info_queue.push(item).unwrap();
+        }
+
+        for _ in 0..image_builders_still_loading.len() {
+            let item = image_builders_still_loading.pop().unwrap();
+            self.image_builders_queue.push(item).unwrap();
+        }
+
+        self.image_info_manager.maintain();
```
Comment (floatingmountain) [inline src/assets/image_asset_manager.rs:112]: Should maintain first i think.

File: src/graphics/material/image.rs [hunk 3]
```diff
@@ -21,13 +21,121 @@ impl Into<wgpu::TextureFormat> for ImageFormat {
...
+        lod_min_clamp: -100.0,
+        lod_max_clamp: 100.0,
+        compare: wgpu::CompareFunction::Undefined,
+    });
+
+    (texture, sampler, texture_extent)
+}
+
+impl ImageBuilder {
+    pub fn build(&self, device: &wgpu::Device, encoder: &mut wgpu::CommandEncoder, image_info: &ImageInfo) -> Image {
```
Comment (floatingmountain) [inline src/graphics/material/image.rs:93]: This looks really nice!

File: src/assets/image_asset_manager.rs (hunk 2 above)
Comment (StarArawn) [reply src/assets/image_asset_manager.rs:112]: I was thinking that, but I wasn't sure..

**Comment by StarArawn** [comment]
Closed in favor of #56

**Pull request closed without merging**